rule_coverage/coverage.py

```diff
@@ -6,116 +6,175 @@
 
+def canonicalize(language):
+  if language in CANONICAL_NAMES:
+    return CANONICAL_NAMES[language]
+  return language
 
-def dump_rule(name, rule, languages, repoAndVersion):
-  if "compatibleLanguages" in rule:
-    for language in rule['compatibleLanguages']:
-      store_rule(name, rule, language, repoAndVersion)
-  else:
-    for language in languages:
-      store_rule(name, rule, language, repoAndVersion)
-
-def store_rule(name, rule, language, repoAndVersion):
-  if language not in rules:
-    print(f"create entry for {language}")
-    rules[language] = {}
-  if '_' in name:
-      name=name[:name.find('_')]
-  if name not in rules[language]:
-    rules[language][name] = repoAndVersion
-
-def dump_rules(repo, version):
+class Coverage:
+  '''Keep and update the coverage DB: lang*ruleId -> analyzer version'''
+  def __init__(self, filename):
+    self.rules = {}
+    if os.path.exists(filename):
+      self.rules = load_json(filename)
+
+  def save_to_file(self, filename):
+    with open(filename, 'w') as outfile:
+      json.dump(self.rules, outfile, indent=2, sort_keys=True)
+
+  def _rule_implemented_for_intermediate_version(self, ruleId, language, repo_and_version):
+    if ruleId not in self.rules[language]:
+      self.rules[language][ruleId] = {'since': repo_and_version, 'until': repo_and_version}
+    elif type(self.rules[language][ruleId]) == dict:
+      self.rules[language][ruleId]['until'] = repo_and_version
+    else:
+      self.rules[language][ruleId] = {'since': self.rules[language][ruleId], 'until': repo_and_version}
+
+  def _rule_implemented_for_last_version(self, ruleId, language, repo_and_version):
+    if ruleId not in self.rules[language]:
+      self.rules[language][ruleId] = repo_and_version
+    elif type(self.rules[language][ruleId]) == dict:
+      self.rules[language][ruleId] = self.rules[language][ruleId]['since']
+
+  def rule_implemented(self, ruleId, language, analyzer, version):
+    repo_and_version = analyzer + ' ' + version
+    language = canonicalize(language)
+
+    if language not in self.rules:
+      print(f"Create entry for {language}")
+      self.rules[language] = {}
+
+    if version == 'master':
+      self._rule_implemented_for_last_version(ruleId, language, repo_and_version)
+    else:
+      self._rule_implemented_for_intermediate_version(ruleId, language, repo_and_version)
+
+  # analyzer+version uniquely identifies the analyzer and version implementing
+  # the rule for the given languages.
+  # Rule implementations for some languages are spread across multiple repositories
+  # for example sonar-java and sonar-security for Java.
+  # We use analyzer+version to avoid confusion between versions of different analyzers.
+  def add_analyzer_version(self, analyzer, version, implemented_rules_per_language):
+    for language in implemented_rules_per_language:
+      for ruleId in implemented_rules[language]:
+        self.rule_implemented(ruleId, language, analyzer, version)
```

Comment:
```suggestion
      for rule_id in implemented_rules_per_language[language]:
        self.rule_implemented(rule_id , language, analyzer, version)
```

I wonder why tests didn't fail... this isn't covered?

Reply:
There are no automated tests for this script (I am adding them as part of RULEAPI-573), and I admit I haven't try to run the script after this last change.